sources/whippet/src/error.rs

```diff
@@ -0,0 +1,135 @@
+//! Error types and handling for whippet.
+//!
+//! This module defines all error types used throughout the application,
+//! providing structured error handling with context information.
+
+use snafu::Snafu;
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub(crate)))]
+pub enum Error {
+    // Configuration errors
+    #[snafu(display("Failed to initialize logger"))]
+    Logger { source: log::SetLoggerError },
+
+    // Policy parsing errors
+    #[snafu(display("Failed to parse policy content"))]
+    PolicyParse { source: toml::de::Error },
+
+    #[snafu(display("Failed to read additional policies directory '{directory}'"))]
+    ReadPolicyDirectory {
+        directory: String,
+        source: std::io::Error,
+    },
+
+    #[snafu(display("Failed to read directory entries"))]
+    ReadPolicyDirectoryEntry { source: std::io::Error },
+
+    #[snafu(display("Failed to read policy file '{path}'"))]
+    ReadPolicyFile {
+        path: String,
+        source: std::io::Error,
+    },
+
+    #[snafu(display("User context includes connect rule"))]
+    UserContextWithConnectRule,
+
+    #[snafu(display("User '{username}' not found"))]
+    UserNotFound { username: String },
+
+    #[snafu(display("Failed to lookup user '{username}'"))]
+    UserLookupFailed {
+        username: String,
+        source: nix::errno::Errno,
+    },
+
+    // Broker errors
+    #[snafu(display("Failed to build D-Bus broker proxy"))]
+    DbusBrokerProxyBuild { source: zbus::Error },
+
+    #[snafu(display("Failed to send policy and socket to the broker"))]
+    DbusBrokerAddListener { source: zbus::Error },
+
+    #[snafu(display("Failed to parse object path for add listener call"))]
+    ParseObjectPath { source: zvariant::Error },
+
+    #[snafu(display("Failed to notify systemd"))]
+    SystemdNotification { source: std::io::Error },
+
+    #[snafu(display("Failed to clone controller socket handle"))]
+    ControllerClone { source: std::io::Error },
+
+    // Socket errors
+    #[snafu(display("Failed to retrieve LISTEN_PID environment variable"))]
+    MissingListenPid { source: std::env::VarError },
+
+    #[snafu(display("Failed to retrieve LISTEN_FDS environment variable"))]
+    MissingListenFds { source: std::env::VarError },
+
+    #[snafu(display("Failed to retrieve NOTIFY_SOCKET environment variable"))]
+    MissingNotifySocket { source: std::env::VarError },
+
+    #[snafu(display("LISTEN_PID mismatch, '{expected}', found {found}"))]
+    UnexpectedPid { expected: String, found: String },
+
+    #[snafu(display("Expected at most 1 file descriptor, foudn {found}"))]
+    UnexpectedFileDescriptorCount { found: u32 },
+
+    #[snafu(display("Invalid file descriptor count: {count}"))]
+    InvalidFileDescriptorCount {
+        count: String,
+        source: std::num::ParseIntError,
+    },
+
+    #[snafu(display("Failed to get FD flags"))]
+    GetFDFlags { source: nix::errno::Errno },
```

Comment:
nit: might also be helpful to log which file descriptor this fails for
```suggestion
    #[snafu(display("Failed to get file descriptor flags"))]
    GetFdFlags { source: nix::errno::Errno },
```